kmod/core/core.c

```diff
@@ -341,27 +357,43 @@ static int kpatch_apply_patch(void *data)
 			hash_del_rcu(&func->node);
 		} while_for_each_linked_func();
 
-		return -EBUSY;
+		ret = -EBUSY;
+		goto err;
 	}
 
-	/* run any user-defined load hooks */
+	/* run any user-defined post-patch callbacks */
 	list_for_each_entry(object, &kpmod->objects, list) {
-		if (!kpatch_object_linked(object))
+		if (!kpatch_object_linked(object) ||
+		    !object->post_patch_callback)
 			continue;
-		list_for_each_entry(hook, &object->hooks_load, list)
-			(*hook->hook)();
-	}
 
+		(*object->post_patch_callback)(object);
+	}
 
 	return 0;
+err:
+	kpatch_state_finish(KPATCH_STATE_FAILURE);
```

Comment:
Better to do this kpatch_state_finish() before the goto, because otherwise it'll be done twice in the NMI activeness safety check failure case.

Reply:
Ah, that's right.